modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/twostep/GridReduceQueryExecutor.java

```diff
@@ -610,6 +613,51 @@ else if (QueryUtils.wasCancelled(err))
         }
     }
 
+    /**
+     * Creates a mapping of node -> expected segments to scan on this particular node.
+     *
+     * @param qry Query to create mapping for.
+     * @param mapping Nodes to partition mapping.
+     * @return Mapping of node to segments.
+     */
+    private Map<ClusterNode, Integer> createNodeToSegmentsCountMapping(GridCacheTwoStepQuery qry, ReducePartitionMapResult mapping) {
+        Map<ClusterNode, Integer> res = new HashMap<>();
+
+        Collection<ClusterNode> nodes = mapping.nodes();
+
+        if (qry.explain() || qry.isReplicatedOnly()) {
+            for (ClusterNode node : nodes) {
+                Integer prev = res.put(node, 1);
+
+                assert prev == null;
+            }
+
+            return res;
+        }
+
+        final int segments = mapper.findFirstPartitioned(qry.cacheIds()).config().getQueryParallelism();
+
+        for (ClusterNode node : nodes) {
+            Map<ClusterNode, IntArray> partsMapping = mapping.queryPartitionsMap();
```

Comment:
escape analysis of course will help here, but you can explicitly move **partsMapping** within the loop scope.